File: backend/app/schemas/experiments.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ExperimentStatus = Literal["draft", "running", "completed", "failed"]


class ExperimentCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    dataset_id: uuid.UUID
    description: str | None = None
    model_provider: str = "deterministic_local"
    model_name: str = "deterministic-extractive-answer-v1"
    embedding_model: str = "deterministic-hash-embedding-384"
    prompt_template_id: uuid.UUID | None = None
    retrieval_strategy: str = "vector_similarity"
    top_k: int = Field(default=5, ge=1, le=50)
    temperature: float = Field(default=0.0, ge=0.0, le=2.0)
    metadata: dict[str, Any] = Field(default_factory=dict)


class ExperimentRead(ExperimentCreate):
    id: uuid.UUID
    status: ExperimentStatus
    created_at: datetime
    updated_at: datetime


class ExperimentRunResponse(BaseModel):
    experiment: ExperimentRead
    examples_run: int
    model_responses_created: int
    evaluations_created: int


class ExperimentAggregateResults(BaseModel):
    experiment_id: uuid.UUID
    status: str
    example_count: int
    avg_correctness: float | None
    avg_groundedness: float | None
    avg_citation_precision: float | None
    avg_citation_recall: float | None
    avg_retrieval_precision: float | None
    avg_retrieval_recall: float | None
    refusal_accuracy: float | None
    hallucination_rate: float | None
    failure_type_counts: dict[str, int]
    avg_latency_ms: float | None
    total_estimated_cost: float | None

File: frontend/types/experiments.ts
export type Experiment = {
  id: string;
  name: string;
  dataset_id: string;
  description?: string | null;
  model_provider: string;
  model_name: string;
  embedding_model: string;
  prompt_template_id?: string | null;
  retrieval_strategy: string;
  top_k: number;
  temperature: number;
  status: "draft" | "running" | "completed" | "failed" | string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type ExperimentResults = {
  experiment_id: string;
  status: string;
  example_count: number;
  avg_correctness?: number | null;
  avg_groundedness?: number | null;
  avg_citation_precision?: number | null;
  avg_citation_recall?: number | null;
  avg_retrieval_precision?: number | null;
  avg_retrieval_recall?: number | null;
  refusal_accuracy?: number | null;
  hallucination_rate?: number | null;
  failure_type_counts: Record<string, number>;
  avg_latency_ms?: number | null;
  total_estimated_cost?: number | null;
};

export type ExperimentResponseRow = {
  response_id: string;
  qa_example_id: string;
  question: string;
  expected_answerability: string;
  model_answerability: string;
  answer_text: string;
  correctness_score?: number | null;
  groundedness_score?: number | null;
  hallucination_flag?: boolean | null;
  failure_type?: string | null;
  created_at: string;
};

export type ExperimentFailureRow = ExperimentResponseRow & {
  failure_reason: string;
};

export type ExperimentReport = {
  experiment: Experiment;
  results: ExperimentResults;
  failure_examples: ExperimentFailureRow[];
  markdown: string;
  generated_at: string;
  disclaimer: string;
  metadata: Record<string, unknown>;
};
